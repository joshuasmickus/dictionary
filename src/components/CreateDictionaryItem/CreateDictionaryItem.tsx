import { Form, Formik } from 'formik';
import * as React from 'react';
import { Redirect } from 'react-router-dom';
import { toast } from 'react-toastify';

import { IDictionaryItem, IDictionaryName, IDictionaryState } from '../../models/common';
import { createDictionaryItem } from '../../state/dictionary/dictionaryRoutines';

import { validate } from './CreateDictionaryItem.validate';

export interface ICreateDictionaryItemProps {
  createDictionaryItemRequest: (values: IDictionaryItem) => void;
  items: IDictionaryItem[];
  list: IDictionaryName[];
  match: any;
};
interface ICreateDictionaryItemState {
  redirect: boolean;
};

export class CreateDictionaryItem extends React.Component<ICreateDictionaryItemProps, ICreateDictionaryItemState> {
  public state = {
    redirect: false
  };

  public render() {
    const { id: dictionaryId } = this.props.match.params;
    const filteredList = this.props.list.filter((dictionary: IDictionaryName) => dictionary.id === dictionaryId);

    if (this.state.redirect || filteredList.length === 0) {
      return <Redirect to={`/edit-dictionary/${dictionaryId}`} />;
    }

    const { name: dictionaryName } = filteredList[0];

    return (
      <Formik
        onSubmit={this.handleSubmit}
        initialValues={{
          from: '',
          to: ''
        }}
        validate={(values) => validate(this.props.items, values)}
      >
        {({
          handleChange,
          handleBlur,
          touched,
          errors,
        }) => (
          <Form>
            <h1>Create dictionary item for {dictionaryName}</h1>
            <label>
              <span>From:</span>
              <input
                name="from"
                required={true}
                type="text"
                onChange={handleChange}
                onBlur={handleBlur}
              />
              {touched.from && errors.from && (<span className="error">{errors.from}</span>)}
            </label>
            <label>
              <span>To:</span>
              <input
                name="to"
                required={true}
                type="text"
                onChange={handleChange}
                onBlur={handleBlur}
              />
              {touched.to && errors.to && (<span  className="error">{errors.to}</span>)}
            </label>
            <label>
              <span>&nbsp;</span>
              <button type="submit">Create dictionary item</button>
            </label>
        </Form>
        )}
      </Formik>
    );
  }

  private handleSubmit = (values: IDictionaryItem) => {
    const { id: dictionaryId } = this.props.match.params;

     this.props.createDictionaryItemRequest({ dictionaryId, ...values });

     toast(`Dictionary item '${values.from} / ${values.to}' created!`);

     this.setState({
      redirect: true
     });
  }
}

export const mapStateToProps = ({ dictionary }: IDictionaryState) => ({
  items: dictionary.items,
  list: dictionary.list
});

export const mapDispatchToProps = (dispatch: (action: any) => void) => ({
  createDictionaryItemRequest: (values: IDictionaryItem) => dispatch(createDictionaryItem.request(values))
});
