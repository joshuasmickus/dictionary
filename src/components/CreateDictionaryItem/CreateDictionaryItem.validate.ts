import { IDictionaryItem } from '../../models/common';

interface IDictionaryErrors {
  from?: string;
  to?: string;
};

export const validate = (items: IDictionaryItem[], values: any) => {
  const errors: IDictionaryErrors = {};

  if (values.from === values.to) {
    errors.from = 'From and to cannot be the same';
    errors.to = 'To and from cannot be the same';
  }

  if (values.from === '') {
    errors.from = 'Enter a from value';
  }

  if (values.to === '') {
    errors.to = 'Enter a to value';
  }

  items.forEach((item: IDictionaryItem) => {
    if(item.from === values.from) {
      errors.from = 'Duplicate from key';
    }

    items.forEach((itemToCheck: IDictionaryItem) => {
      if (values.from === item.from && values.to === itemToCheck.to) {
        errors.from = 'From is a clone of another dictionary item';
        errors.to = 'To is a clone of another dictionary item';
      }
    });
  });

  return errors;
};
