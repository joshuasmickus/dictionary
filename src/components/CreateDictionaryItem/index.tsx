import { connect } from 'react-redux';

import {
  CreateDictionaryItem,
  mapDispatchToProps,
  mapStateToProps
} from './CreateDictionaryItem';

import './CreateDictionaryItem.css';

export default connect(mapStateToProps, mapDispatchToProps)(CreateDictionaryItem);
