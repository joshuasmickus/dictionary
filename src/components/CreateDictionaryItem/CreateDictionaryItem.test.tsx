import { createMemoryHistory } from 'history'
import * as React from 'react';
import { Router } from 'react-router-dom';
import { render } from 'react-testing-library';

import { CreateDictionaryItem } from './CreateDictionaryItem';

const id = 'someid';
const requiredProps = {
  createDictionaryItemRequest: jest.fn(),
  items: [],
  list: [{ name: 'something', id }],
  match: { params: { id }}
};

describe('Given a CreateDictionaryItem component', () => {
  describe('when it is rendered', () => {
    it('should match the snapshot', () => {
      const { container } = render(<Router history={createMemoryHistory()}><CreateDictionaryItem {...requiredProps} /></Router>);

      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
