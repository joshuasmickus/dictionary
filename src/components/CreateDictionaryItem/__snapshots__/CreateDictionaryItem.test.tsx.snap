// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a CreateDictionaryItem component when it is rendered should match the snapshot 1`] = `
<form>
  <h1>
    Create dictionary item for 
    something
  </h1>
  <label>
    <span>
      From:
    </span>
    <input
      name="from"
      required=""
      type="text"
    />
  </label>
  <label>
    <span>
      To:
    </span>
    <input
      name="to"
      required=""
      type="text"
    />
  </label>
  <label>
    <span>
      
    </span>
    <button
      type="submit"
    >
      Create dictionary item
    </button>
  </label>
</form>
`;
